### PR TITLE
Add project registration feature

### PR DESCRIPTION
**Pull request by Larusso**
## Description

This patch allows to register gradle sub projects, package files or directories as snyk projects. The usecase is to have more control over the invocation of the snyk cli when dealing with a gradle multi project. The `--all-projects` or `--all-sub-projects` flag of the `snyk` cli has issues if one want to control the project name or if one needs to set custom tags/properties per project.

This is where project registration comes into play. It allows to configure a whole multiproject from the root build.gradle file. A potential configuration could look like this:

```groovy
snyk {
    projectLifecycle = "production"
    projectEnvironment = "internal"
    projectBusinessCriticality = "high"
    projectTags = ["team": "someTeam", "component": "lib"]
    strategies.set(['monitor_check', "report_check", 'monitor_publish'])
    failOn = "all"
    severityThreshold = "high"

    registerProject(file("paket.dependencies")) {
        projectTags.put("platform", ".Net")
        projectEnvironment = "distributed"
    }

    registerProject(project.findProject(":Shared")) {
        projectTags.put("platform", "universal")
    }

    registerProject(project.subprojects.findAll {it.path.startsWith(":android:")}) {
        projectTags.put("platform", "android")
        projectTags.put("component", "android-lib")
        projectEnvironment = "mobile"
    }
}
```

For this to work I created a handfull of new classes and interfaces. The biggest impact is the split between a root snyk extension and a normal snyk extension. Internally I create new extensions for each registered project. The extensions will be added either on the to be registered sub project or will be added in root with a unique name. I don't want allow that these extensions contain the methods to register projects on their own again to limit the fail vector of misconfiguration. So the root snyk extension which is added when applying the plugin will be the only extension which contains the newly added register methods.

For each registered project the new `ProjectRegistrationHandler` will create an extension and the three base tasks (`test`, `monitor`, `report`). Depending on the type of project to be added (gradle subproject or file) the tasks will be either added on the sub projects or the root project with a unique name based on the file name.

The conventions will cascade up to the root extensions only for a few exceptions.

Example: the newly registered `snykTest` task for the sub project `sub1` will be created in `sub1` and named `snykTest`.
Or fully qualified `:sub1:snykTest`.
The property token will be bound to the property token of registered `SnykPluginExtension` named `snyk` in `sub1`. The property token of the `SnykPluginExtension` in `sub1` will be bound to the `snyk` extension in the root project.

```
:sub1:snykTest#token -> :sub1:snyk#token -> :snyk#token
```

Obviously this can be all reconfigured. For this I added overloads to all register methods to pass either an `Action` object or a configuration closure.

Because of the name mangling needed to add multiple snykTest tasks to the root project it might become hard to configure specific properties for a given task only. This is the reason why I added new task accessors to the extension.

```groovy
snyk {
    snykMonitor {    // configure the snykMonitor task in root
        projectTags = ...
    }

    registerProject(project.findProject("sub")) {
       snykMonitor { // configure the snykMonitor task in :sub project
           token = ...
       }
    }

    registerProject(file('project.properties')) {
       snykMonitor { // configure the snykMonitor task for `project.properties` project file
           token = ...
       }
    }
}
```

The extensions add accessors for
* snykTest
* snykMonitor
* snykReport

Changes
=======

* ![ADD] snyk project register support

[NEW]:      https://resources.atlas.wooga.com/icons/icon_new.svg "New"
[ADD]:      https://resources.atlas.wooga.com/icons/icon_add.svg "Add"
[IMPROVE]:  https://resources.atlas.wooga.com/icons/icon_improve.svg "Improve"
[CHANGE]:   https://resources.atlas.wooga.com/icons/icon_change.svg "Change"
[FIX]:      https://resources.atlas.wooga.com/icons/icon_fix.svg "Fix"
[UPDATE]:   https://resources.atlas.wooga.com/icons/icon_update.svg "Update"

[BREAK]:    https://resources.atlas.wooga.com/icons/icon_break.svg "Remove"
[REMOVE]:   https://resources.atlas.wooga.com/icons/icon_remove.svg "Remove"
[IOS]:      https://resources.atlas.wooga.com/icons/icon_iOS.svg "iOS"
[ANDROID]:  https://resources.atlas.wooga.com/icons/icon_android.svg "Android"
[WEBGL]:    https://resources.atlas.wooga.com/icons/icon_webGL.svg "WebGL"
[GRADLE]:   https://resources.atlas.wooga.com/icons/icon_gradle.svg "GRADLE"
[UNITY]:    https://resources.atlas.wooga.com/icons/icon_unity.svg "Unity"
[LINUX]:    https://resources.atlas.wooga.com/icons/icon_linux.svg "Linux"
[WIN]:      https://resources.atlas.wooga.com/icons/icon_windows.svg "Windows"
[MACOS]:    https://resources.atlas.wooga.com/icons/icon_iOS.svg "macOS"

### CODE REVIEW

File: src/main/groovy/wooga/gradle/snyk/SnykPlugin.groovy [hunk 1]
```diff
@@ -66,16 +80,143 @@ class SnykPlugin implements Plugin<Project> {
...
+
+    SnykPluginExtension registerProject(File projectFile, SnykRootPluginExtension parentExtension) {
+        def relativeProjectFile = project.relativePath(projectFile)
+        def projectName = project.name + ":" + relativeProjectFile
+
+        if (projectFile.isFile()) {
+            logger.info("register snyk project file: ${projectFile.absolutePath}")
+        } else {
+            logger.info("register snyk project at path: ${projectFile.absolutePath}")
+        }
```
Comment (Azurelol) [inline src/main/groovy/wooga/gradle/snyk/SnykPlugin.groovy:113]: Should it not check if the given project file does not exist?

Comment (Larusso) [reply src/main/groovy/wooga/gradle/snyk/SnykPlugin.groovy:113]: Yes correct. That as well. I saw already during testing that `isFile` will be false if the file does not exist. I will add the check and a test.

Comment (Azurelol) [reply src/main/groovy/wooga/gradle/snyk/SnykPlugin.groovy:113]: Got it!

**Comment by Larusso** [comment]
@Azurelol I added the file check and a matching test. In the future I would like to be able to support lazy file pointer so a task could generate them. But I don't see this as a must at the moment.

**Pull request merged**